packages/dx-react-grid/docs/guides/optimizing-performance.md

```diff
@@ -0,0 +1,27 @@
+# React Components - Optimizing Performance
+
+Internally, Reactive Components use React [guidelines](https://reactjs.org/docs/optimizing-performance.html) for optimizing performance. However, a customer component configuration is just as important an internal configuration for performance optimization. You can avoid big performance problems if you follow some easy rules.
+
+## Use Immutable Data Structures
+
+React Controls do not change data passed through a state and uses memoization and [React.PureComponents](https://reactjs.org/docs/react-api.html#reactpurecomponent). A `React.PureComponent` compares old and new state values using a simple comparison. In this case, if you update a state object field, React does not update the component because it compares two references to the same object. Immutable data structures help to solve this problem.
+
+Keep the Control’s state management plugins as separate components and update their properties independently to optimize performance. This allows the Control to avoid rendering unchanged UI elements.
+
+In this case, you can use a state management library like [Redux](https://redux.js.org/) to work with independent components’ states.
+
+The following demo shows how to work with immutable data using Redux, [react-redux](https://github.com/reduxjs/react-redux) and [seamless-immutable](https://github.com/rtfeldman/seamless-immutable) libraries:
+
+.embedded-demo({ "path": "optimizing-performance/grid-immutable", "showThemeSelector": true })
+
+## Avoid Constantly Code Inside Render Methods
+
+It is known that all code inside render methods will be called after each component's updates. We do not recommend creating new functions and variables in this place. To override default templates, you should create your own templates outside render methods. The following example demonstrates the right way to override default the [React Scheduler](https://devexpress.github.io/devextreme-reactive/react/scheduler) cell templates:
```

Comment:
"to override default **the** cell templates" -> "to override default cell templates"